File: go/openpgp/openpgp.go
// SPDX-FileCopyrightText: 2026 Blackcat Informatics® Inc. <paudley@blackcatinformatics.ca>
// SPDX-License-Identifier: MIT OR Apache-2.0

// Package openpgp is a minimal reader for Ed25519 armored public keys, used by
// `gts extract-key` (§9.2).
//
// It mirrors the Python gts.openpgp reference: it parses only the unencrypted
// armored public-key certificates GPG emits for Ed25519 (OpenPGP algorithm 22)
// keys, extracting the raw 32-byte key and computing the v4 fingerprint so GTS
// tooling can show the embedded transport key without shelling out to gpg.
// Everything else (other algorithms, encrypted secret keys, v5/v6 packets) is
// rejected with a clear error.
package openpgp

import (
	"bytes"
	"crypto/sha1" //nolint:gosec // RFC 4880 §12.2 mandates SHA-1 for v4 fingerprints; not a security primitive here.
	"encoding/base64"
	"encoding/binary"
	"errors"
	"fmt"
	"strings"
)

const ed25519Algo = 22

// ed25519OID is the curve OID GPG writes for the Ed25519 signing curve
// (1.3.6.1.4.1.11591.15.1).
var ed25519OID = []byte{0x2b, 0x06, 0x01, 0x04, 0x01, 0xda, 0x47, 0x0f, 0x01}

// TransportKey is a parsed Ed25519 public key: the raw 32-byte key plus its
// uppercase 40-hex-character OpenPGP v4 fingerprint.
type TransportKey struct {
	RawPublic   []byte
	Fingerprint string
}

// stripArmor decodes the packet bytes from an ASCII-armored OpenPGP block.
func stripArmor(text string) ([]byte, error) {
	lines := strings.Split(text, "\n")
	start, end := -1, -1
	for i, l := range lines {
		if start == -1 && strings.HasPrefix(l, "-----BEGIN PGP") {
			start = i
		} else if start != -1 && i > start && strings.HasPrefix(l, "-----END PGP") {
			end = i
			break
		}
	}
	if start == -1 {
		return nil, errors.New("missing armor BEGIN line")
	}
	if end == -1 {
		return nil, errors.New("missing armor END line")
	}

	idx := start + 1
	// Skip optional armor headers (Comment, Version, …) up to the blank line.
	for idx < end && strings.TrimSpace(lines[idx]) != "" {
		if strings.Contains(lines[idx], ":") {
			idx++
		} else {
			break
		}
	}

	var body strings.Builder
	for idx < end {
		line := strings.TrimRight(lines[idx], "\r")
		if strings.HasPrefix(line, "=") {
			break // CRC-24 checksum line — end of the base64 body.
		}
		body.WriteString(line)
		idx++
	}
	if body.Len() == 0 {
		return nil, errors.New("empty armor body")
	}
	out, err := base64.StdEncoding.DecodeString(body.String())
	if err != nil {
		return nil, errors.New("invalid base64 armor body")
	}
	return out, nil
}

// readMPI reads an OpenPGP multi-precision integer; returns (bytes, nextOffset).
func readMPI(data []byte, offset int) ([]byte, int, error) {
	if offset+2 > len(data) {
		return nil, 0, errors.New("truncated MPI length")
	}
	bits := int(binary.BigEndian.Uint16(data[offset : offset+2]))
	length := (bits + 7) / 8
	end := offset + 2 + length
	if end > len(data) {
		return nil, 0, errors.New("truncated MPI payload")
	}
	return data[offset+2 : end], end, nil
}

// nextPacket parses one OpenPGP packet; returns (tag, body, nextOffset).
// Supports both old- and new-format headers.
func nextPacket(data []byte, offset int) (int, []byte, int, error) {
	if offset >= len(data) {
		return 0, nil, 0, errors.New("truncated packet header")
	}
	header := data[offset]
	if header&0x80 == 0 {
		return 0, nil, 0, errors.New("invalid packet tag octet")
	}

	var tag, length int
	if header&0x40 != 0 {
		// New-format packet.
		tag = int(header & 0x3f)
		offset++
		if offset >= len(data) {
			return 0, nil, 0, errors.New("truncated new-format length octet")
		}
		lo := data[offset]
		switch {
		case lo < 192:
			length = int(lo)
			offset++
		case lo < 224:
			if offset+1 >= len(data) {
				return 0, nil, 0, errors.New("truncated new-format 2-octet length")
			}
			length = (int(lo)-192)<<8 + int(data[offset+1]) + 192
			offset += 2
		case lo == 255:
			if offset+4 >= len(data) {
				return 0, nil, 0, errors.New("truncated new-format 4-octet length")
			}
			length = int(binary.BigEndian.Uint32(data[offset+1 : offset+5]))
			offset += 5
		default:
			return 0, nil, 0, errors.New("partial body lengths are not supported")
		}
	} else {
		// Old-format packet.
		tag = int((header >> 2) & 0x0f)
		lengthType := header & 0x03
		offset++
		switch lengthType {
		case 0:
			if offset >= len(data) {
				return 0, nil, 0, errors.New("truncated old-format length octet")
			}
			length = int(data[offset])
			offset++
		case 1:
			if offset+1 >= len(data) {
				return 0, nil, 0, errors.New("truncated old-format 2-octet length")
			}
			length = int(binary.BigEndian.Uint16(data[offset : offset+2]))
			offset += 2
		case 2:
			if offset+3 >= len(data) {
				return 0, nil, 0, errors.New("truncated old-format 4-octet length")
			}
			length = int(binary.BigEndian.Uint32(data[offset : offset+4]))
			offset += 4
		default:
			return 0, nil, 0, errors.New("indeterminate-length packets are not supported")
		}
	}

	end := offset + length
	if end > len(data) {
		return 0, nil, 0, errors.New("packet body exceeds input")
	}
	return tag, data[offset:end], end, nil
}

// packet is one parsed OpenPGP packet.
type packet struct {
	tag  int
	body []byte
}

// iterPackets returns every (tag, body) packet in the de-armored data.
func iterPackets(data []byte) ([]packet, error) {
	var packets []packet
	offset := 0
	for offset < len(data) {
		tag, body, next, err := nextPacket(data, offset)
		if err != nil {
			return nil, err
		}
		packets = append(packets, packet{tag: tag, body: body})
		offset = next
	}
	return packets, nil
}

// parseEd25519PublicMaterial parses the OID and raw key from a v4 public-key
// packet body; returns (rawPublicKey, endOffsetOfPublicMaterial).
func parseEd25519PublicMaterial(body []byte) ([]byte, int, error) {
	if len(body) < 6 || body[0] != 4 {
		return nil, 0, errors.New("only OpenPGP v4 public keys are supported")
	}
	if body[5] != ed25519Algo {
		return nil, 0, fmt.Errorf("unsupported public-key algorithm %d", body[5])
	}
	offset := 6
	if offset >= len(body) {
		return nil, 0, errors.New("truncated public-key packet")
	}
	oidLen := int(body[offset])
	offset++
	if offset+oidLen > len(body) {
		return nil, 0, errors.New("truncated OID")
	}
	oid := body[offset : offset+oidLen]
	offset += oidLen
	if !bytes.Equal(oid, ed25519OID) {
		return nil, 0, fmt.Errorf("unsupported curve OID %x", oid)
	}

	mpi, end, err := readMPI(body, offset)
	if err != nil {
		return nil, 0, err
	}
	// GPG encodes the Ed25519 public key as a 33-byte MPI (0x40 || 32-byte key);
	// a bare 32-byte MPI is also valid when the high bit is clear.
	switch len(mpi) {
	case 33:
		return append([]byte(nil), mpi[1:]...), end, nil
	case 32:
		return append([]byte(nil), mpi...), end, nil
	default:
		return nil, 0, fmt.Errorf("unexpected Ed25519 public MPI length %d", len(mpi))
	}
}

// fingerprint computes the OpenPGP v4 fingerprint of a public-key packet body:
// SHA-1(0x99 || u16-be(len(body)) || body), uppercased.
func fingerprint(pubKeyBody []byte) string {
	h := sha1.New() //nolint:gosec // RFC 4880 v4 fingerprint construction.
	h.Write([]byte{0x99})
	var lenBuf [2]byte
	binary.BigEndian.PutUint16(lenBuf[:], uint16(len(pubKeyBody)))
	h.Write(lenBuf[:])
	h.Write(pubKeyBody)
	return fmt.Sprintf("%X", h.Sum(nil))
}

// ParseTransportKey parses an armored OpenPGP certificate into its raw Ed25519
// key and v4 fingerprint. It accepts either a public-key certificate (tag 6) or
// an unencrypted secret-key block (tag 5); the fingerprint always covers only
// the public material.
func ParseTransportKey(armored string) (TransportKey, error) {
	data, err := stripArmor(armored)
	if err != nil {
		return TransportKey{}, err
	}
	packets, err := iterPackets(data)
	if err != nil {
		return TransportKey{}, err
	}
	for _, p := range packets {
		var raw, pubBody []byte
		switch p.tag {
		case 6:
			raw, _, err = parseEd25519PublicMaterial(p.body)
			if err != nil {
				return TransportKey{}, err
			}
			pubBody = p.body
		case 5:
			var end int
			raw, end, err = parseEd25519PublicMaterial(p.body)
			if err != nil {
				return TransportKey{}, err
			}
			pubBody = p.body[:end]
		default:
			continue
		}
		return TransportKey{RawPublic: raw, Fingerprint: fingerprint(pubBody)}, nil
	}
	return TransportKey{}, errors.New("no public-key packet found")
}

File: kotlin/src/main/kotlin/ca/blackcatinformatics/gts/OpenPgp.kt
// SPDX-FileCopyrightText: 2026 Blackcat Informatics Inc. <paudley@blackcatinformatics.ca>
// SPDX-License-Identifier: MIT OR Apache-2.0

package ca.blackcatinformatics.gts

import java.security.MessageDigest
import java.util.Base64

private const val ED25519_ALGO = 22
private val ED25519_OID = byteArrayOf(0x2b, 0x06, 0x01, 0x04, 0x01, 0xda.toByte(), 0x47, 0x0f, 0x01)
private val EMOJI =
    arrayOf(
        "🐵", "🐶", "🐺", "🦊", "🐱", "🦁", "🐯", "🐴",
        "🦄", "🦓", "🦌", "🐮", "🐷", "🐗", "🐭", "🐹",
        "🐰", "🐻", "🐼", "🐨", "🐸", "🐲", "🐔", "🐧",
        "🦆", "🦅", "🦉", "🦇", "🐢", "🐍", "🦎", "🐊",
        "🐳", "🐬", "🐟", "🐠", "🐡", "🦈", "🐙", "🦑",
        "🦀", "🦞", "🦐", "🦋", "🐌", "🐞", "🐝", "🐜",
        "🦂", "🍎", "🍐", "🍊", "🍋", "🍌", "🍉", "🍇",
        "🍓", "🍒", "🍍", "🥝", "🍑", "🥥", "🥕", "🌽",
    )

data class TransportKey(val rawPublic: ByteArray, val fingerprint: String) {
    override fun equals(other: Any?): Boolean =
        other is TransportKey && rawPublic.contentEquals(other.rawPublic) && fingerprint == other.fingerprint

    override fun hashCode(): Int = 31 * rawPublic.contentHashCode() + fingerprint.hashCode()
}

fun parseTransportKey(armored: String): TransportKey {
    val packets = iterPackets(stripArmor(armored))
    for (packet in packets) {
        when (packet.tag) {
            6 -> {
                val parsed = parseEd25519PublicMaterial(packet.body)
                return TransportKey(parsed.raw, fingerprint(packet.body))
            }
            5 -> {
                val parsed = parseEd25519PublicMaterial(packet.body)
                return TransportKey(parsed.raw, fingerprint(packet.body.copyOfRange(0, parsed.end)))
            }
        }
    }
    error("no public-key packet found")
}

fun formatFingerprint(fingerprint: String): String {
    val compact = fingerprint.filterNot { it.isWhitespace() }.uppercase()
    if (compact.isEmpty() || compact.any { it !in "0123456789ABCDEF" }) return fingerprint
    return compact.chunked(4).joinToString(" ")
}

fun emojihash(data: ByteArray, length: Int = 11): String {
    val wanted = length.coerceAtLeast(1)
    val nbytes = (wanted * 6 + 7) / 8
    val digest = blake3_256(data).copyOfRange(0, nbytes)
    val out = mutableListOf<Int>()
    var acc = 0L
    var bits = 0
    for (b in digest) {
        acc = (acc shl 8) or (b.toLong() and 0xff)
        bits += 8
        while (bits >= 6 && out.size < wanted) {
            bits -= 6
            out += ((acc shr bits) and 0x3f).toInt()
        }
        acc = acc and ((1L shl bits) - 1)
    }
    return out.take(wanted).joinToString(" ") { EMOJI[it] }
}

private data class Packet(val tag: Int, val body: ByteArray)

private data class PublicMaterial(val raw: ByteArray, val end: Int)

private fun stripArmor(text: String): ByteArray {
    val lines = text.split('\n')
    val start = lines.indexOfFirst { it.startsWith("-----BEGIN PGP") }
    require(start >= 0) { "missing armor BEGIN line" }
    val end = lines.drop(start + 1).indexOfFirst { it.startsWith("-----END PGP") }.let {
        if (it < 0) -1 else it + start + 1
    }
    require(end >= 0) { "missing armor END line" }
    var idx = start + 1
    while (idx < end && lines[idx].trim().isNotEmpty() && ":" in lines[idx]) idx++
    val body = StringBuilder()
    while (idx < end) {
        val line = lines[idx].trimEnd('\r')
        if (line.startsWith("=")) break
        body.append(line)
        idx++
    }
    require(body.isNotEmpty()) { "empty armor body" }
    return try {
        Base64.getDecoder().decode(body.toString())
    } catch (_: IllegalArgumentException) {
        throw IllegalArgumentException("invalid base64 armor body")
    }
}

private fun iterPackets(data: ByteArray): List<Packet> {
    val packets = mutableListOf<Packet>()
    var offset = 0
    while (offset < data.size) {
        val next = nextPacket(data, offset)
        packets += Packet(next.tag, next.body)
        offset = next.next
    }
    return packets
}

private data class PacketRead(val tag: Int, val body: ByteArray, val next: Int)

private fun nextPacket(data: ByteArray, initialOffset: Int): PacketRead {
    var offset = initialOffset
    require(offset < data.size) { "truncated packet header" }
    val header = data[offset].toInt() and 0xff
    require(header and 0x80 != 0) { "invalid packet tag octet" }
    var tag: Int
    var length: Int
    if (header and 0x40 != 0) {
        tag = header and 0x3f
        offset++
        require(offset < data.size) { "truncated new-format length octet" }
        val lo = data[offset].toInt() and 0xff
        when {
            lo < 192 -> {
                length = lo
                offset++
            }
            lo < 224 -> {
                require(offset + 1 < data.size) { "truncated new-format 2-octet length" }
                length = ((lo - 192) shl 8) + (data[offset + 1].toInt() and 0xff) + 192
                offset += 2
            }
            lo == 255 -> {
                require(offset + 4 < data.size) { "truncated new-format 4-octet length" }
                length =
                    ((data[offset + 1].toInt() and 0xff) shl 24) or
                        ((data[offset + 2].toInt() and 0xff) shl 16) or
                        ((data[offset + 3].toInt() and 0xff) shl 8) or
                        (data[offset + 4].toInt() and 0xff)
                offset += 5
            }
            else -> error("partial body lengths are not supported")
        }
    } else {
        tag = (header shr 2) and 0x0f
        val lengthType = header and 0x03
        offset++
        when (lengthType) {
            0 -> {
                require(offset < data.size) { "truncated old-format length octet" }
                length = data[offset].toInt() and 0xff
                offset++
            }
            1 -> {
                require(offset + 1 < data.size) { "truncated old-format 2-octet length" }
                length = ((data[offset].toInt() and 0xff) shl 8) or (data[offset + 1].toInt() and 0xff)
                offset += 2
            }
            2 -> {
                require(offset + 3 < data.size) { "truncated old-format 4-octet length" }
                length =
                    ((data[offset].toInt() and 0xff) shl 24) or
                        ((data[offset + 1].toInt() and 0xff) shl 16) or
                        ((data[offset + 2].toInt() and 0xff) shl 8) or
                        (data[offset + 3].toInt() and 0xff)
                offset += 4
            }
            else -> error("indeterminate-length packets are not supported")
        }
    }
    val end = offset + length
    require(end <= data.size) { "packet body exceeds input" }
    return PacketRead(tag, data.copyOfRange(offset, end), end)
}

private fun parseEd25519PublicMaterial(body: ByteArray): PublicMaterial {
    require(body.size >= 6 && body[0].toInt() == 4) { "only OpenPGP v4 public keys are supported" }
    require((body[5].toInt() and 0xff) == ED25519_ALGO) { "unsupported public-key algorithm ${body[5].toInt() and 0xff}" }
    var offset = 6
    require(offset < body.size) { "truncated public-key packet" }
    val oidLen = body[offset].toInt() and 0xff
    offset++
    require(offset + oidLen <= body.size) { "truncated OID" }
    val oid = body.copyOfRange(offset, offset + oidLen)
    offset += oidLen
    require(oid.contentEquals(ED25519_OID)) { "unsupported curve OID ${hex(oid)}" }
    val mpi = readMpi(body, offset)
    val raw =
        when (mpi.bytes.size) {
            33 -> mpi.bytes.copyOfRange(1, 33)
            32 -> mpi.bytes
            else -> error("unexpected Ed25519 public MPI length ${mpi.bytes.size}")
        }
    return PublicMaterial(raw, mpi.next)
}

private data class Mpi(val bytes: ByteArray, val next: Int)

private fun readMpi(data: ByteArray, offset: Int): Mpi {
    require(offset + 2 <= data.size) { "truncated MPI length" }
    val bits = ((data[offset].toInt() and 0xff) shl 8) or (data[offset + 1].toInt() and 0xff)
    val length = (bits + 7) / 8
    val end = offset + 2 + length
    require(end <= data.size) { "truncated MPI payload" }
    return Mpi(data.copyOfRange(offset + 2, end), end)
}

private fun fingerprint(pubKeyBody: ByteArray): String {
    val digest = MessageDigest.getInstance("SHA-1")
    digest.update(0x99.toByte())
    digest.update(((pubKeyBody.size ushr 8) and 0xff).toByte())
    digest.update((pubKeyBody.size and 0xff).toByte())
    digest.update(pubKeyBody)
    return hex(digest.digest()).uppercase()
}
